### ISSUE TITLE
Reconnection problems when connection goes down

### ISSUE THREAD

**Issue by vict2019** [Azure/azure-event-hubs-node#178]
I've downloaded the latest package and I don't see the Unauthorized access error anymore but it looks like a problem still happens when trying to reconnect.  I'm seeing the following error message:

The connection was inactive for more than the allowed 60000 milliseconds and is closed by container...

And then the receivers don't reconnect and eventually get the message that something is busted.  Here are more logs:
```
azure:event-hubs:error [connection-1] Error (context.connection.error) occurred on the amqp connection: c {
  azure:event-hubs:error   value:
  azure:event-hubs:error    [ Typed { type: [Object], value: 'amqp:connection:forced' },
  azure:event-hubs:error      Typed {
  azure:event-hubs:error        type: [Object],
  azure:event-hubs:error        value: 'The connection was inactive for more than the allowed 60000 milliseconds and is closed by container \'LinkTracker\'. TrackingId:b8f8f89478fc4cc3beea90f3e89ef207_G5, SystemTracker:gateway5, Timestamp:2019-01-06T10:35:43' },
  azure:event-hubs:error      Typed { type: [Object], value: null } ] } +45m
  azure:event-hubs:error [connection-1] state: { wasConnectionCloseCalled: false,
  azure:event-hubs:error   numSenders: 0,
  azure:event-hubs:error   numReceivers: 24 } +2ms
  azure:event-hubs:error [connection-1] connection.close() was not called from the sdk and there were some sender or receiver links or both. We should reconnect. +1ms
  azure:event-hubs:error [connection-1] calling detached on receiver '4d8a97bf-4886-46d9-9b5d-0c0bafaf83b7' with address 'victoryiothubdev/ConsumerGroups/eventpublisher/Partitions/0'. +301ms
  rhea-promise:error [connection-1] The receiver is open ? -> false +45m
  azure:event-hubs:error [connection-1] calling detached on receiver 'b49b6c50-2464-4f30-8543-b0a252409804' with address 'victoryiothubdev/ConsumerGroups/eventpublisher/Partitions/1'. +2ms
  rhea-promise:error [connection-1] The receiver is open ? -> false +1ms
  azure:event-hubs:error [connection-1] calling detached on receiver '759a93a3-ed16-46f6-b1fd-88b2f560bb2a' with address 'victoryiothubdev/ConsumerGroups/eventpublisher/Partitions/2'. +1ms
  rhea-promise:error [connection-1] The receiver is open ? -> false +0ms
  azure:event-hubs:error [connection-1] calling detached on receiver '2731d83e-0d34-4e29-a3d2-1622cbe23788' with address 'victoryiothubdev/ConsumerGroups/eventpublisher/Partitions/3'. +1ms
  rhea-promise:error [connection-1] The receiver is open ? -> false +1ms
  azure:event-hubs:error [connection-1] calling detached on receiver 'b49b6c50-2464-4f30-8543-b0a252409804' with address 'victoryiothubdev/ConsumerGroups/eventpublisher/Partitions/1'. +0ms
  rhea-promise:error [connection-1] The receiver is open ? -> false +1ms
  azure:event-hubs:error [connection-1] calling detached on receiver 'b49b6c50-2464-4f30-8543-b0a252409804' with address 'victoryiothubdev/ConsumerGroups/eventpublisher/Partitions/1'. +1ms
  rhea-promise:error [connection-1] The receiver is open ? -> false +0ms
  azure:event-hubs:error [connection-1] calling detached on receiver 'b49b6c50-2464-4f30-8543-b0a252409804' with address 'victoryiothubdev/ConsumerGroups/eventpublisher/Partitions/1'. +1ms
  rhea-promise:error [connection-1] The receiver is open ? -> false +1ms
  azure:event-hubs:error [connection-1] calling detached on receiver '4d8a97bf-4886-46d9-9b5d-0c0bafaf83b7' with address 'victoryiothubdev/ConsumerGroups/eventpublisher/Partitions/0'. +0ms
  rhea-promise:error [connection-1] The receiver is open ? -> false +1ms
  azure:event-hubs:error [connection-1] calling detached on receiver '4d8a97bf-4886-46d9-9b5d-0c0bafaf83b7' with address 'victoryiothubdev/ConsumerGroups/eventpublisher/Partitions/0'. +1ms
  rhea-promise:error [connection-1] The receiver is open ? -> false +0ms
  azure:event-hubs:error [connection-1] calling detached on receiver '4d8a97bf-4886-46d9-9b5d-0c0bafaf83b7' with address 'victoryiothubdev/ConsumerGroups/eventpublisher/Partitions/0'. +0ms
  rhea-promise:error [connection-1] The receiver is open ? -> false +1ms
  azure:event-hubs:error [connection-1] calling detached on receiver '2731d83e-0d34-4e29-a3d2-1622cbe23788' with address 'victoryiothubdev/ConsumerGroups/eventpublisher/Partitions/3'. +1ms
  rhea-promise:error [connection-1] The receiver is open ? -> false +0ms
  azure:event-hubs:error [connection-1] calling detached on receiver '2731d83e-0d34-4e29-a3d2-1622cbe23788' with address 'victoryiothubdev/ConsumerGroups/eventpublisher/Partitions/3'. +1ms
  rhea-promise:error [connection-1] The receiver is open ? -> false +1ms
  azure:event-hubs:error [connection-1] calling detached on receiver '4d8a97bf-4886-46d9-9b5d-0c0bafaf83b7' with address 'victoryiothubdev/ConsumerGroups/eventpublisher/Partitions/0'. +0ms
  rhea-promise:error [connection-1] The receiver is open ? -> false +1ms
  azure:event-hubs:error [connection-1] calling detached on receiver '4d8a97bf-4886-46d9-9b5d-0c0bafaf83b7' with address 'victoryiothubdev/ConsumerGroups/eventpublisher/Partitions/0'. +1ms
  rhea-promise:error [connection-1] The receiver is open ? -> false +0ms
  azure:event-hubs:error [connection-1] calling detached on receiver '2731d83e-0d34-4e29-a3d2-1622cbe23788' with address 'victoryiothubdev/ConsumerGroups/eventpublisher/Partitions/3'. +7ms
  rhea-promise:error [connection-1] The receiver is open ? -> false +8ms
  azure:event-hubs:error [connection-1] calling detached on receiver '2731d83e-0d34-4e29-a3d2-1622cbe23788' with address 'victoryiothubdev/ConsumerGroups/eventpublisher/Partitions/3'. +1ms
  rhea-promise:error [connection-1] The receiver is open ? -> false +0ms
  azure:event-hubs:error [connection-1] calling detached on receiver '2731d83e-0d34-4e29-a3d2-1622cbe23788' with address 'victoryiothubdev/ConsumerGroups/eventpublisher/Partitions/3'. +1ms
  rhea-promise:error [connection-1] The receiver is open ? -> false +1ms
  azure:event-hubs:error [connection-1] calling detached on receiver '4d8a97bf-4886-46d9-9b5d-0c0bafaf83b7' with address 'victoryiothubdev/ConsumerGroups/eventpublisher/Partitions/0'. +0ms
  rhea-promise:error [connection-1] The receiver is open ? -> false +1ms
  azure:event-hubs:error [connection-1] calling detached on receiver '4d8a97bf-4886-46d9-9b5d-0c0bafaf83b7' with address 'victoryiothubdev/ConsumerGroups/eventpublisher/Partitions/0'. +1ms
  rhea-promise:error [connection-1] The receiver is open ? -> false +0ms
  azure:event-hubs:error [connection-1] calling detached on receiver '4d8a97bf-4886-46d9-9b5d-0c0bafaf83b7' with address 'victoryiothubdev/ConsumerGroups/eventpublisher/Partitions/0'. +1ms
  rhea-promise:error [connection-1] The receiver is open ? -> false +1ms
  azure:event-hubs:error [connection-1] calling detached on receiver '759a93a3-ed16-46f6-b1fd-88b2f560bb2a' with address 'victoryiothubdev/ConsumerGroups/eventpublisher/Partitions/2'. +0ms
  rhea-promise:error [connection-1] The receiver is open ? -> false +1ms
  azure:event-hubs:error [connection-1] calling detached on receiver 'b49b6c50-2464-4f30-8543-b0a252409804' with address 'victoryiothubdev/ConsumerGroups/eventpublisher/Partitions/1'. +1ms
  rhea-promise:error [connection-1] The receiver is open ? -> false +1ms
  azure:event-hubs:error [connection-1] calling detached on receiver '2731d83e-0d34-4e29-a3d2-1622cbe23788' with address 'victoryiothubdev/ConsumerGroups/eventpublisher/Partitions/3'. +1ms
  rhea-promise:error [connection-1] The receiver is open ? -> false +1ms
  azure:event-hubs:error [connection-1] calling detached on receiver '2731d83e-0d34-4e29-a3d2-1622cbe23788' with address 'victoryiothubdev/ConsumerGroups/eventpublisher/Partitions/3'. +1ms
  rhea-promise:error [connection-1] The receiver is open ? -> false +1ms
  rhea-promise:error [connection-1] The session is open ? -> false +0ms
  rhea-promise:error [connection-1] The session is open ? -> false +0ms
  rhea-promise:error [connection-1] The session is open ? -> false +0ms
  rhea-promise:error [connection-1] The session is open ? -> false +6ms
  rhea-promise:error [connection-1] The session is open ? -> false +0ms
  rhea-promise:error [connection-1] The session is open ? -> false +1ms
  rhea-promise:error [connection-1] The session is open ? -> false +0ms
  rhea-promise:error [connection-1] The session is open ? -> false +0ms
  rhea-promise:error [connection-1] The session is open ? -> false +1ms
  rhea-promise:error [connection-1] The session is open ? -> false +0ms
  rhea-promise:error [connection-1] The session is open ? -> false +1ms
  rhea-promise:error [connection-1] The session is open ? -> false +0ms
  rhea-promise:error [connection-1] The session is open ? -> false +0ms
  rhea-promise:error [connection-1] The session is open ? -> false +1ms
  rhea-promise:error [connection-1] The session is open ? -> false +0ms
  rhea-promise:error [connection-1] The session is open ? -> false +0ms
  rhea-promise:error [connection-1] The session is open ? -> false +1ms
  rhea-promise:error [connection-1] The session is open ? -> false +0ms
  rhea-promise:error [connection-1] The session is open ? -> false +0ms
  rhea-promise:error [connection-1] The session is open ? -> false +1ms
  rhea-promise:error [connection-1] The session is open ? -> false +0ms
  rhea-promise:error [connection-1] The session is open ? -> false +0ms
  rhea-promise:error [connection-1] The session is open ? -> false +0ms
  rhea-promise:error [connection-1] The session is open ? -> false +1ms
  azure:event-hubs:error [connection-1] Something is busted. State of Receiver '4d8a97bf-4886-46d9-9b5d-0c0bafaf83b7' with address 'victoryiothubdev/ConsumerGroups/eventpublisher/Partitions/0' is: { wasCloseInitiated: true,
  azure:event-hubs:error   receiverError: c { value: [ [Object], [Object], [Object] ] },
  azure:event-hubs:error   _receiver:
  azure:event-hubs:error    Receiver {
  azure:event-hubs:error      domain: null,
  azure:event-hubs:error      _events:
  azure:event-hubs:error       { message: [Function],
  azure:event-hubs:error         receiver_error: [Function],
  azure:event-hubs:error         receiver_close: [Function] },
  azure:event-hubs:error      _eventsCount: 3,
  azure:event-hubs:error      _maxListeners: undefined,
  azure:event-hubs:error      actionInitiated: 0,
  azure:event-hubs:error      type: 'receiver',
  azure:event-hubs:error      _session:
  azure:event-hubs:error       Session {
  azure:event-hubs:error         domain: null,
  azure:event-hubs:error         _events: [Object],
  azure:event-hubs:error         _eventsCount: 2,
  azure:event-hubs:error         _maxListeners: undefined,
  azure:event-hubs:error         actionInitiated: 0,
  azure:event-hubs:error         _connection: [Object],
  azure:event-hubs:error         _session: [Object] },
  azure:event-hubs:error      _link:
  azure:event-hubs:error       Receiver {
  azure:event-hubs:error         session: [Object],
  azure:event-hubs:error         connection: [Object],
  azure:event-hubs:error         name: 'e102e390-0e1f-478f-9295-a0125c1221d7',
  azure:event-hubs:error         options: [Object],
  azure:event-hubs:error         state: [Object],
  azure:event-hubs:error         issue_flow: false,
  azure:event-hubs:error         local: [Object],
  azure:event-hubs:error         remote: [Object],
  azure:event-hubs:error         delivery_count: 17686,
  azure:event-hubs:error         credit: 814,
  azure:event-hubs:error         observers: [Object],
  azure:event-hubs:error         drain: false,
  azure:event-hubs:error         _events: [Object],
  azure:event-hubs:error         _eventsCount: 7 },
  azure:event-hubs:error      linkOptions:
  azure:event-hubs:error       { name: 'e102e390-0e1f-478f-9295-a0125c1221d7',
  azure:event-hubs:error         autoaccept: true,
  azure:event-hubs:error         source: [Object],
  azure:event-hubs:error         credit_window: 1000,
  azure:event-hubs:error         onMessage: [Function],
  azure:event-hubs:error         onError: [Function],
  azure:event-hubs:error         onClose: [Function],
  azure:event-hubs:error         onSessionError: [Function],
  azure:event-hubs:error         onSessionClose: [Function] } } } +14ms
  azure:event-hubs:error [connection-1] Something is busted. State of Receiver 'b49b6c50-2464-4f30-8543-b0a252409804' with address 'victoryiothubdev/ConsumerGroups/eventpublisher/Partitions/1' is: { wasCloseInitiated: true,
  azure:event-hubs:error   receiverError: c { value: [ [Object], [Object], [Object] ] },

```

**Comment by KyotoFox** [comment]
Is anything being done to address this issue?

**Comment by ramya-rao-a** [comment]
Thanks for reporting this @vict2019 and @KyotoFox, we are looking into this and will get back as soon as we have something to share.

**Comment by ramya-rao-a** [comment]
@vict2019 and @KyotoFox Just giving an update here that we have found the root cause as why we are not retrying the connection and resuming the receiver
- Host not found is a non retryable error. Unfortunately Nodejs gives the same `ENOTFOUND` error for both host not found and the network being down scenarios. Therefore, we were treating the network being down as a non retryable error as well, and so recevier didnt get resumed.
- [rhea](https://github.com/amqp/rhea) is the library we use to create the AMQP connection and AMQP links for senders/receivers. The indicator for suggesting if the link was explicitly closed by us or due to external factors had a bug. This is what caused the logs you see for `Something is busted`

The second issue has been now fixed in rhea. We are in the process of fixing the first issue.
We are targeting sometime next week to publish an update with the fixes.

Thanks for your patience

**Comment by ramya-rao-a** [comment]
@vict2019, @KyotoFox Thanks for your patience

We have just released a new version 2.0.0 for the Event Hubs library which includes the fix for the connection related problems being discussed in this issue.

We will be re-connecting and resume receiving of messages when the network goes down and then comes back up

Please keep in mind that there is a major version update in this release. This is because of a few breaking changes for which you may have to react to and make some changes in your code. Please see the [changelog](https://github.com/Azure/azure-sdk-for-js/blob/master/packages/%40azure/eventhubs/client/changelog.md#2019-03-26-200) for the list of breaking changes and bug fixes.
